examples/vision/mirnet.py

```diff
@@ -0,0 +1,516 @@
+    return layers.Add()([input_tensor, conv])
+
+
+"""
+### MIRNet Model
+"""
+
+
+def recursive_residual_group(input_tensor, num_mrb, channels):
+    conv1 = layers.Conv2D(channels, kernel_size=(3, 3), padding="same")(input_tensor)
+    for _ in range(num_mrb):
+        conv1 = multi_scale_residual_block(conv1, channels)
+    conv2 = layers.Conv2D(channels, kernel_size=(3, 3), padding="same")(conv1)
+    return layers.Add()([conv2, input_tensor])
+
+
+def mirnet_model(num_rrg, num_mrb, channels):
+    input_tensor = keras.Input(shape=[None, None, 3])
+    x1 = layers.Conv2D(channels, kernel_size=(3, 3), padding="same")(input_tensor)
+    for _ in range(num_rrg):
+        x1 = recursive_residual_group(x1, num_mrb, channels)
+    conv = layers.Conv2D(3, kernel_size=(3, 3), padding="same")(x1)
+    output_tensor = layers.Add()([input_tensor, conv])
+    return keras.Model(input_tensor, output_tensor)
+
+
+model = mirnet_model(num_rrg=3, num_mrb=2, channels=64)
+
+"""
+## Training
+
+- We train MIRNet using **Charbonnier Loss** as the loss function and **Adam Optimizer**
+with a learning rate of `1e-4`.
+
+- We use **Peak Signal Noise Ratio** or PSNR as a metric which is an expression for the
+ratio between the maximum possible value (power) of a signal and the power of distorting
+noise that affects the quality of its representation.
+"""
+
+
+def charbonnier_loss(y_true, y_pred):
+    return tf.reduce_mean(tf.sqrt(tf.square(y_true - y_pred) + tf.square(1e-3)))
+
+
+def peak_signal_noise_ratio(y_true, y_pred):
+    return tf.image.psnr(y_pred, y_true, max_val=255.0)
+
+
+optimizer = keras.optimizers.Adam(learning_rate=1e-4)
+model.compile(
+    optimizer=optimizer, loss=charbonnier_loss, metrics=[peak_signal_noise_ratio]
+)
+
+history = model.fit(
+    train_dataset,
+    validation_data=val_dataset,
+    epochs=50,
+    callbacks=[
+        keras.callbacks.ReduceLROnPlateau(
+            monitor="val_peak_signal_noise_ratio",
+            factor=0.5,
+            patience=5,
+            verbose=1,
+            min_delta=1e-7,
+            mode="max",
+        )
+    ],
+)
+
+plt.plot(history.history["loss"], label="train_loss")
+plt.plot(history.history["val_loss"], label="val_loss")
+plt.xlabel("Epochs")
+plt.ylabel("Loss")
+plt.title("Train and Validation Losses Over Epochs", fontsize=14)
+plt.legend()
+plt.grid()
+plt.show()
+
+
+plt.plot(history.history["peak_signal_noise_ratio"], label="train_psnr")
+plt.plot(history.history["val_peak_signal_noise_ratio"], label="val_psnr")
+plt.xlabel("Epochs")
+plt.ylabel("PSNR")
+plt.title("Train and Validation PSNR Over Epochs", fontsize=14)
+plt.legend()
+plt.grid()
+plt.show()
+
+"""
+## Inference
+
+For inference, we need the dimensions of the input image to be multiples of 4. Hence we
+resize the input image to the closest possible dimension that is acceptable by the model
+before inference.
+"""
+
+
+def closest_number(n, m):
```

Comment:
Please add a one-line description of this function

Reply:
Cleaned up the inference code and removed the unnecessary `closest_number` function.